Allow using a function for user-provided cover.

FlorentBecker opened the pull request:
This makes it unnecessary to duplicate the metadata / cfg such as author, title etc when using a custom cover.

FlorentBecker committed:
Allow using a function for user-provided cover.

This makes it unnecessary to duplicate the metadata / cfg

mayconfmelo commented:
Hello! Thank you so much for your contribution! I've been wanting to implement something like this to pave the way for cover templates, but I've never thought of such an elegant approach like yours.

The variables would be accessed like:

```typst
(title, subtitle, date, authors, volume, cfg) => { title }
```

or:

```typst
(..meta) => { meta.pos().at(0) }
```

right?!

FlorentBecker commented:
I'm not sure how to add an example to the manual, but using this option, you'd do something like:

```typst
#let mkCover(title, subtitle, date, authors, volume, cfg) = {
let the_authors = authors.join(", ", last: ", and ")
align(horizon+center)[
  #title,
  #linebreak()
  a book by #the_authors
  #linebreak()
  writ #date
]
}

#show: book.with(
     title: [My life],
     authors: ([Me], [Myself], [I]),
     cover: mkCover,
     cfg: cfg,
 )

= First part

this starts the content of the book
```

mayconfmelo commented:
Cool! I like this usage. And based on your clever approach, I have a small idea for a revision and would like your feedback.

What if, instead of explicitly receiving six variables, the function only received one dictionary?

```typst
#let mkCover(meta) = {
  let the_authors = meta.authors.join(", ", last: ", and ")
  align(horizon+center)[
    #meta.title,
    #linebreak()
    a book by #the_authors
    #linebreak()
    writ #meta.date
  ]
}

#show: book.with(
     title: [My life],
     authors: ("Me", "Myself", "I"),
     cover: mkCover,
 )
```

Thus, variables not used by the function are not explicitly declared, but they are still available in the dictionary—in your example, `subtitle`, `volume`, and `cfg` must be declared, but are not used.

What do you think? Are there any drawbacks to this approach?

FlorentBecker commented:
My gut feeling says to have the function accept two arguments, `meta` and `cfg`, as I don't like having `meta.cfg` too much.

mayconfmelo commented on src/components/cover.typ:
```diff
@@ -74,6 +74,10 @@
     ]
   }
+  else if type(cover) == function {
+      cover(title, subtitle, date, authors, volume, cfg)
```
This should return two dictionaries: one with metadata and cfg.

```suggestion
      cover((
        title: title,
        subtitle: subtitle,
        date: date,
        authors: authors,
        volume: volume,
        ),
        cfg
      )
```

mayconfmelo reviewed (changes requested):
Check this out:

FlorentBecker committed:
Change the type of the cover callback

A cover function now goes (meta, cfg) => content where meta is the metadata dictionnary (title, subtitle, date, authors, volume), and cfg the config dictionnary.

Co-authored-by: Maycon F. Melo <30307203+mayconfmelo@users.noreply.github.com>
